servers/src/main/java/tachyon/worker/block/TieredBlockStore.java

```diff
@@ -0,0 +1,432 @@
+ * a BlockMetadataManager to maintain the status of the tiered storage, and a LockManager to
+ * coordinate read/write on the same block.
+ * <p>
+ * This class is thread-safe.
+ */
+public class TieredBlockStore implements BlockStore {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  private final TachyonConf mTachyonConf;
+  private final BlockMetadataManager mMetaManager;
+  private final BlockLockManager mLockManager;
+  private final Allocator mAllocator;
+  private final Evictor mEvictor;
+  private List<BlockStoreEventListener> mBlockStoreEventListeners =
+      new ArrayList<BlockStoreEventListener>();
+
+  /** A readwrite lock for meta data **/
+  private final ReentrantReadWriteLock mEvictionLock = new ReentrantReadWriteLock();
+
+  public TieredBlockStore(TachyonConf tachyonConf) throws IOException {
+    mTachyonConf = Preconditions.checkNotNull(tachyonConf);
+    mMetaManager = BlockMetadataManager.newBlockMetadataManager(mTachyonConf);
+    mLockManager = new BlockLockManager(mMetaManager);
+
+    AllocatorType allocatorType =
+        mTachyonConf.getEnum(Constants.WORKER_ALLOCATE_STRATEGY_TYPE, AllocatorType.DEFAULT);
+    mAllocator = AllocatorFactory.create(allocatorType, mMetaManager);
+
+    EvictorType evictorType =
+        mTachyonConf.getEnum(Constants.WORKER_EVICT_STRATEGY_TYPE, EvictorType.DEFAULT);
+    mEvictor = EvictorFactory.create(evictorType, mMetaManager);
+    if (mEvictor instanceof BlockStoreEventListener) {
+      registerBlockStoreEventListener((BlockStoreEventListener) mEvictor);
+    }
+  }
+
+  @Override
+  public long lockBlock(long userId, long blockId) throws IOException {
+    return mLockManager.lockBlock(userId, blockId, BlockLockType.READ);
+  }
+
+  @Override
+  public void unlockBlock(long lockId) throws IOException {
+    mLockManager.unlockBlock(lockId);
+  }
+
+  @Override
+  public void unlockBlock(long userId, long blockId) throws IOException {
+    mLockManager.unlockBlock(userId, blockId);
+  }
+
+  @Override
+  public BlockWriter getBlockWriter(long userId, long blockId) throws IOException {
+    TempBlockMeta tempBlockMeta = mMetaManager.getTempBlockMeta(blockId);
+    return new LocalFileBlockWriter(tempBlockMeta);
+  }
+
+  @Override
+  public BlockReader getBlockReader(long userId, long blockId, long lockId) throws IOException {
+    mLockManager.validateLockId(userId, blockId, lockId);
+    BlockMeta blockMeta = mMetaManager.getBlockMeta(blockId);
+    return new LocalFileBlockReader(blockMeta);
+  }
+
+  @Override
+  public TempBlockMeta createBlockMeta(long userId, long blockId, BlockStoreLocation location,
+      long initialBlockSize) throws IOException {
+    TempBlockMeta tempBlock;
+    mEvictionLock.readLock().lock();
+    try {
+      tempBlock = createBlockMetaNoLock(userId, blockId, location, initialBlockSize);
+    } finally {
+      mEvictionLock.readLock().unlock();
+    }
+    return tempBlock;
```

Comment:
I think this can be simplified to

```
try {
  return createBlockMetaNoLock(userId, blockId, location, initialBlockSize);
} finally {
  mEvictionLock.readLock().unlock();
}
```